File: src/redactor_mcp/comprehend.py
import logging
import os

import boto3
from botocore.exceptions import (
    ClientError,
    CredentialRetrievalError,
    NoCredentialsError,
    PartialCredentialsError,
    TokenRetrievalError,
)
from mypy_boto3_comprehend import ComprehendClient
from mypy_boto3_comprehend.literals import LanguageCodeType
from mypy_boto3_comprehend.type_defs import PiiEntityTypeDef

logger = logging.getLogger(__name__)

_client: ComprehendClient | None = None


def _get_client() -> ComprehendClient:
    global _client
    if _client is None:
        region = os.environ.get("AWS_REGION") or os.environ.get("AWS_DEFAULT_REGION") or "us-east-1"
        try:
            _client = boto3.client("comprehend", region_name=region)
        except Exception as error:
            # Check if it's a credential error and provide helpful message
            if _is_credential_error(error) or "sso" in str(error).lower():
                logger.error(f"Failed to create Comprehend client due to credential error: {error}")

                # Get the AWS profile for the error message
                profile = os.environ.get("AWS_PROFILE")
                if profile:
                    profile_msg = f"aws sso login --profile {profile}"
                else:
                    profile_msg = "aws sso login --profile <your-profile>"

                raise RuntimeError(
                    f"AWS credentials are invalid or expired. "
                    f"If using SSO, run: {profile_msg}"
                ) from error
            raise
    return _client


def _reset_client() -> None:
    """Clear the cached client to force recreation on next access."""
    global _client
    _client = None
    logger.info("Comprehend client reset due to credential error")


def _is_credential_error(error: Exception) -> bool:
    """Check if the error is credential-related."""
    # Check botocore credential exception types
    if isinstance(
        error,
        (
            NoCredentialsError,
            PartialCredentialsError,
            CredentialRetrievalError,
            TokenRetrievalError,
        ),
    ):
        return True

    # Check ClientError codes for expired/invalid tokens
    if isinstance(error, ClientError):
        error_code = error.response.get("Error", {}).get("Code", "")
        credential_error_codes = {
            "ExpiredTokenException",
            "ExpiredToken",
            "InvalidToken",
            "InvalidClientTokenId",
            "UnrecognizedClientException",
        }
        if error_code in credential_error_codes:
            return True

    return False


def detect_pii_entities(
    text: str, language_code: LanguageCodeType = "en"
) -> list[PiiEntityTypeDef]:
    """Call AWS Comprehend DetectPiiEntities and return the list of entities.

    Automatically retries once with a fresh client if credential errors are detected.
    """
    client = _get_client()

    try:
        response = client.detect_pii_entities(Text=text, LanguageCode=language_code)
        return response["Entities"]
    except Exception as error:
        # If it's a credential error, reset the client and retry once
        if _is_credential_error(error):
            logger.warning(f"Credential error detected: {error}. Resetting client and retrying...")
            _reset_client()

            # Retry with fresh client
            try:
                client = _get_client()
                response = client.detect_pii_entities(Text=text, LanguageCode=language_code)
                logger.info("Retry succeeded with fresh credentials")
                return response["Entities"]
            except Exception as retry_error:
                logger.error(f"Retry failed: {retry_error}")
                raise

        # Not a credential error or retry failed - propagate exception
        raise


def redact_text(text: str, entities: list[PiiEntityTypeDef]) -> str:
    """Replace each entity span with [TYPE]. Processes in reverse offset order."""
    sorted_entities = sorted(
        entities, key=lambda e: e.get("BeginOffset", 0), reverse=True
    )
    result = text
    for entity in sorted_entities:
        begin = entity.get("BeginOffset", 0)
        end = entity.get("EndOffset", 0)
        tag = f"[{entity.get('Type', 'UNKNOWN')}]"
        result = result[:begin] + tag + result[end:]
    return result

File: src/comprehend.ts
import {
  ComprehendClient,
  DetectDominantLanguageCommand,
  DetectPiiEntitiesCommand,
  type DominantLanguage,
  type PiiEntity,
} from "@aws-sdk/client-comprehend";
import { CredentialsProviderError } from "@smithy/property-provider";

// Cached client singleton
let _client: ComprehendClient | null = null;

/**
 * Get AWS region from environment variables with fallback.
 */
function getRegion(): string {
  return (
    process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION || "us-east-1"
  );
}

/**
 * Check if an error is credential-related (matches Python implementation logic).
 */
function isCredentialError(error: unknown): boolean {
  if (!error || typeof error !== "object") return false;

  // Check Smithy credential provider errors (AWS SDK v3)
  if (error instanceof CredentialsProviderError) {
    return true;
  }

  // Check error name for known credential error types
  if ("name" in error) {
    const credentialErrors = new Set([
      "ExpiredTokenException",
      "ExpiredToken",
      "InvalidToken",
      "InvalidClientTokenId",
      "UnrecognizedClientException",
      "CredentialsProviderError",
      "ProviderError",
    ]);
    if (credentialErrors.has(error.name as string)) {
      return true;
    }
  }

  // Check HTTP status codes (401/403 typically auth failures)
  if ("$metadata" in error) {
    const metadata = (error as any).$metadata;
    if (metadata?.httpStatusCode === 401 || metadata?.httpStatusCode === 403) {
      return true;
    }
  }

  // Check error message for SSO/credential keywords
  const errorStr = error.toString().toLowerCase();
  if (
    errorStr.includes("sso") ||
    errorStr.includes("credential") ||
    errorStr.includes("token")
  ) {
    return true;
  }

  return false;
}

/**
 * Get the cached Comprehend client or create a new one.
 * SDK v3 resolves credentials lazily on first API call, so construction never throws.
 */
function getClient(): ComprehendClient {
  if (_client === null) {
    _client = new ComprehendClient({ region: getRegion() });
  }
  return _client;
}

/**
 * Clear the cached client to force recreation on next access.
 */
function resetClient(): void {
  _client = null;
}

/**
 * Call AWS Comprehend DetectDominantLanguage and return the list of detected languages.
 * Automatically retries once with a fresh client if credential errors are detected.
 */
export async function detectLanguage(text: string): Promise<DominantLanguage[]> {
  const client = getClient();

  try {
    const command = new DetectDominantLanguageCommand({ Text: text });
    const response = await client.send(command);
    return response.Languages || [];
  } catch (error) {
    // If it's a credential error, reset the client and retry once
    if (isCredentialError(error)) {
      resetClient();

      // Retry with fresh client
      try {
        const freshClient = getClient();
        const command = new DetectDominantLanguageCommand({ Text: text });
        const response = await freshClient.send(command);
        return response.Languages || [];
      } catch (retryError) {
        throw retryError;
      }
    }

    // Not a credential error or retry failed - propagate exception
    throw error;
  }
}

/**
 * Call AWS Comprehend DetectPiiEntities and return the list of entities.
 * Automatically retries once with a fresh client if credential errors are detected.
 */
export async function detectPiiEntities(
  text: string,
  languageCode: string = "en"
): Promise<PiiEntity[]> {
  const client = getClient();
  const langCode = languageCode || "en";

  try {
    const command = new DetectPiiEntitiesCommand({
      Text: text,
      LanguageCode: langCode as "en",
    });
    const response = await client.send(command);
    return response.Entities || [];
  } catch (error) {
    // If it's a credential error, reset the client and retry once
    if (isCredentialError(error)) {
      resetClient();

      const freshClient = getClient();
      const command = new DetectPiiEntitiesCommand({
        Text: text,
        LanguageCode: languageCode as "en",
      });
      const response = await freshClient.send(command);
      return response.Entities || [];
    }

    // Not a credential error - propagate exception
    throw error;
  }
}

/**
 * Replace each entity span with [TYPE]. Processes in reverse offset order
 * to maintain correct positions during replacement.
 */
export function redactText(text: string, entities: PiiEntity[]): string {
  // Sort entities by offset in reverse order
  const sortedEntities = [...entities].sort(
    (a, b) => (b.BeginOffset ?? 0) - (a.BeginOffset ?? 0)
  );

  let result = text;
  for (const entity of sortedEntities) {
    const begin = entity.BeginOffset ?? 0;
    const end = entity.EndOffset ?? 0;
    const tag = `[${entity.Type ?? "UNKNOWN"}]`;
    result = result.slice(0, begin) + tag + result.slice(end);
  }

  return result;
}
